File: clients/android/app/src/main/kotlin/com/lextures/android/core/config/SchoolCodeLogic.kt
package com.lextures.android.core.config

/**
 * Validates school codes the same way as the marketing site (`www/src/lib/school-code.ts`).
 * Special case: `local` routes the mobile app to the local API for development.
 */
object SchoolCodeLogic {
    const val SELF_LEARNER_API_BASE = "https://self.lextures.com"
    const val LOCAL_API_BASE = "http://127.0.0.1:8080"
    const val TENANT_HOST_SUFFIX = "lextures.com"

    private val pattern = Regex("^[a-z][a-z0-9]*(?:-[a-z0-9]+)*$")

    private val reserved = setOf(
        "admin",
        "api",
        "app",
        "default",
        "demo",
        "login",
        "magic-link",
        "mfa",
        "self",
        "signup",
        "www",
    )

    fun normalize(raw: String): String = raw.trim().lowercase()

    /**
     * Returns an Android string resource name for the validation error, or null if valid.
     * `local` is always accepted (dev shortcut).
     */
    fun errorKey(code: String): String? {
        val normalized = normalize(code)
        if (normalized.isEmpty()) return "auth_getStarted_schoolCodeErrorEmpty"
        if (normalized == "local") return null
        if (normalized.length < 2) return "auth_getStarted_schoolCodeErrorLengthMin"
        if (normalized.length > 32) return "auth_getStarted_schoolCodeErrorLengthMax"
        if (!pattern.matches(normalized)) return "auth_getStarted_schoolCodeErrorFormat"
        if (normalized in reserved) return "auth_getStarted_schoolCodeErrorReserved"
        return null
    }

    fun isValid(code: String): Boolean = errorKey(code) == null

    fun apiBaseUrl(schoolCode: String): String {
        val normalized = normalize(schoolCode)
        if (normalized == "local") return LOCAL_API_BASE
        return "https://$normalized.$TENANT_HOST_SUFFIX"
    }

    fun previewHost(schoolCode: String): String {
        val normalized = normalize(schoolCode)
        if (normalized.isEmpty()) return "your-school.$TENANT_HOST_SUFFIX"
        if (normalized == "local") return "127.0.0.1:8080"
        return "$normalized.$TENANT_HOST_SUFFIX"
    }
}

File: clients/ios/Lextures/Core/Config/SchoolCodeLogic.swift
import Foundation

/// Validates school codes the same way as the marketing site (`www/src/lib/school-code.ts`).
/// Special case: `local` routes the mobile app to the local API for development.
enum SchoolCodeLogic {
    static let homeschoolAPIBase = "https://self.lextures.com"
    static let localAPIBase = "http://127.0.0.1:8080"
    static let tenantHostSuffix = "lextures.com"

    private static let pattern = makeRegex(#"^[a-z][a-z0-9]*(?:-[a-z0-9]+)*$"#)

    private static func makeRegex(_ pattern: String) -> NSRegularExpression {
        guard let regex = try? NSRegularExpression(pattern: pattern) else {
            preconditionFailure("Invalid school-code regex: \(pattern)")
        }
        return regex
    }

    private static let reserved: Set<String> = [
        "admin",
        "api",
        "app",
        "default",
        "demo",
        "login",
        "magic-link",
        "mfa",
        "self",
        "signup",
        "www",
    ]

    static func normalize(_ raw: String) -> String {
        raw.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    }

    /// Returns a localization key for the validation error, or `nil` if valid.
    /// `local` is always accepted (dev shortcut).
    static func errorKey(for code: String) -> String? {
        let normalized = normalize(code)
        if normalized.isEmpty {
            return "auth.getStarted.schoolCodeErrorEmpty"
        }
        if normalized == "local" {
            return nil
        }
        if normalized.count < 2 {
            return "auth.getStarted.schoolCodeErrorLengthMin"
        }
        if normalized.count > 32 {
            return "auth.getStarted.schoolCodeErrorLengthMax"
        }
        let range = NSRange(normalized.startIndex..<normalized.endIndex, in: normalized)
        if pattern.firstMatch(in: normalized, options: [], range: range) == nil {
            return "auth.getStarted.schoolCodeErrorFormat"
        }
        if reserved.contains(normalized) {
            return "auth.getStarted.schoolCodeErrorReserved"
        }
        return nil
    }

    static func isValid(_ code: String) -> Bool {
        errorKey(for: code) == nil
    }

    static func apiBaseURL(schoolCode: String) -> String {
        let normalized = normalize(schoolCode)
        if normalized == "local" {
            return localAPIBase
        }
        return "https://\(normalized).\(tenantHostSuffix)"
    }

    static func previewHost(schoolCode: String) -> String {
        let normalized = normalize(schoolCode)
        if normalized.isEmpty {
            return "your-school.\(tenantHostSuffix)"
        }
        if normalized == "local" {
            return "127.0.0.1:8080"
        }
        return "\(normalized).\(tenantHostSuffix)"
    }
}
